### PR TITLE
Add value completion for properties

### PR DESCRIPTION
**Pull request by evie-lau**
Add value completion for properties in server.env and bootstrap.properties files #51, #52
The `ServerPropertyValues` class contains the HashMaps for properties to valid values, and utilizes the `EquivalentProperties` for bootstrap properties to grab the same values as server.env properties.

Property key completion: 
image

Value completion:
image

TODO: ensure all properties with a set of valid values are added to `ServerPropertyValues`
Will investigate support for default values/completion separately in another PR

### CODE REVIEW

**Comment by evie-lau** [comment]
Question: for properties that can define directory paths such as these, should we have the autocomplete suggest the default value?
```
wlp.user.dir=The directory of the usr folder. The default is `${wlp.install.dir}/usr`.
shared.app.dir=The directory of shared applications. The default is `${wlp.user.dir}/shared/apps`.
shared.config.dir=The directory of shared configuration files. The default is `${wlp.user.dir}/shared/config`.
shared.resource.dir=The directory of shared resource files. The default is `${wlp.user.dir}/shared/resources`.
```

**Comment by cherylking** [comment]
> Question: for properties that can define directory paths such as these, should we have the autocomplete suggest the default value?
> 
> ```
> wlp.user.dir=The directory of the usr folder. The default is `${wlp.install.dir}/usr`.
> shared.app.dir=The directory of shared applications. The default is `${wlp.user.dir}/shared/apps`.
> shared.config.dir=The directory of shared configuration files. The default is `${wlp.user.dir}/shared/config`.
> shared.resource.dir=The directory of shared resource files. The default is `${wlp.user.dir}/shared/resources`.
> ```

I think you should ask this in the devex guild slack channel.

File: liberty-ls/src/main/java/io/openliberty/tools/langserver/utils/ServerPropertyValues.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
+package io.openliberty.tools.langserver.utils;
```
Comment (cherylking) [inline liberty-ls/src/main/java/io/openliberty/tools/langserver/utils/ServerPropertyValues.java:1]: Add copyright.

File: liberty-ls/src/main/resources/ServerEnv.properties [hunk 2]
```diff
@@ -9,10 +9,6 @@ WLP_OUTPUT_DIR=The directory that contains output files for defined servers. Thi
 WLP_DEBUG_ADDRESS=The port to use for running the server in debug mode. The default value is `7777`.
 WLP_DEBUG_SUSPEND=Whether to suspend the JVM on startup or not. This variable can be set to `y` to suspend the JVM on startup until a debugger attaches, or set to `n` to start up without waiting for a debugger to attach. The default value is `y`.
 WLP_DEBUG_REMOTE=Whether to allow remote debugging or not. This variable can be set to `y` to allow remote debugging. By default, this value is not defined, which does not allow remote debugging on newer JDK/JREs.
-WLP_OUTPUT_DIR=The directory that contains output files for defined servers. This directory must have both read and write permissions for the user or users who start servers. By default, the server output logs and work area are stored at the `%WLP_USER_DIR%/servers/serverName` location alongside configuration and applications. If this variable is set, the output logs and work area are stored at the `%WLP_OUTPUT_DIR%/serverName location`.
```
Comment (cherylking) [inline liberty-ls/src/main/resources/ServerEnv.properties:12]: Why are these being removed?

File: liberty-ls/src/main/java/io/openliberty/tools/langserver/utils/ServerPropertyValues.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+
+import io.openliberty.tools.langserver.ls.LibertyTextDocument;
+
+/**
+ * Maps property keys with their list of valid values for server.env and bootstrap.properties
+ */
+public class ServerPropertyValues {
```
Comment (cherylking) [inline liberty-ls/src/main/java/io/openliberty/tools/langserver/utils/ServerPropertyValues.java:12]: What about properties that have a certain data type, like it should be an integer? Is there anything we will do to flag incorrect values for those?

File: liberty-ls/src/main/resources/ServerEnv.properties (hunk 2 above)
Comment (evie-lau) [reply liberty-ls/src/main/resources/ServerEnv.properties:12]: These are duplicate entries

File: liberty-ls/src/main/java/io/openliberty/tools/langserver/utils/ServerPropertyValues.java (hunk 3 above)
Comment (cherylking) [reply liberty-ls/src/main/java/io/openliberty/tools/langserver/utils/ServerPropertyValues.java:12]: Ok, if we don't have an issue open for validating values, please go ahead and open one.

File: liberty-ls/src/main/java/io/openliberty/tools/langserver/model/propertiesfile/PropertiesEntryInstance.java [hunk 4]
```diff
@@ -70,11 +70,9 @@ public CompletableFuture<Hover> getHover(Position position) {
     }
 
     public CompletableFuture<List<CompletionItem>> getCompletions(Position position) {
-        // TODO: may have to revisit this for proper completions. doing only keys for now
-        // if (isOnEntryKey(position)) {
-            return propertyKeyInstance.getCompletions(position);
-        // } else {
-        //     return propertyValueInstance.getCompletions(position);
-        // }
+        if (propertyValueInstance.toString() != null) {
```
[the remark this answers is not in the document]
Comment (evie-lau) [reply liberty-ls/src/main/java/io/openliberty/tools/langserver/model/propertiesfile/PropertiesEntryInstance.java]: Added the position check

File: liberty-ls/src/main/java/io/openliberty/tools/langserver/utils/Messages.java [hunk 5]
```diff
@@ -47,14 +51,19 @@ public static String getPropDescription(String key) {
...
+            } else if (key.contains(".")) { // bootstrap property
+                if (EquivalentProperties.hasEquivalentProperty(key)) { // bootstrap property has equivalent server.env property
+                    message = serverenvMessages.getString(EquivalentProperties.getEquivalentProperty(key));
+                } else {
+                    message = bootstrapMessages.getString(key);
+                }
             }
+        } catch (MissingResourceException e) {
+            // Caught to avoid unnecessary console output when hovering over invalid keys
+            LOGGER.fine("Can't find property description for: " + key);
```
Comment (cherylking) [inline liberty-ls/src/main/java/io/openliberty/tools/langserver/utils/Messages.java:66]: Suggested rewording: "No property description found for: "

**Review by cherylking — CHANGES_REQUESTED** [review]

**Pull request merged**